research/conditional/moe_layers/continuous_moe.py

```diff
@@ -53,32 +53,30 @@ def __post_init__(self):
         self.original_group_size = self.group_size
 
     def forward(self, x):
-        x = self.rearrange_for_grouping(x)
+        x = self.reshape_into_groups(x)
         merge_weights, emit_weights = self.get_merge_and_emit_weights(x)
         x = self.merge_map_emit(x, merge_weights, emit_weights)
         x = self.reshape_into_original(x)
         return x * (self.group_size / self.original_group_size)
 
-    def rearrange_for_grouping(self, x):
+    def reshape_into_groups(self, x):
         """
         :param x: normal input tensor of shape (B, S, dmodel)
-        :return: x transposed so that the dimension to group over is second last
+        :return: x of shape (free_dimension, aggr_dimension // group_size, group_size , dmodel)
```

Comment:
maybe:

Reshape code so the axis to group over is on position 1, and then group the tensor over said axis.
Example: 

- If we group tokens from different sequences in a batch (sparsity = 0), we need to put the batch dimension to position 1.
- If we group tokens within one sequence, the dimension to group over is already on position 1, hence we leave it as is.